File: AspNetMvcUnleashed/0672329980_source/Chapter21/VB/UnleashedBlog/Models/BlogService.vb
Public Class BlogService
    Inherits BlogServiceBase

    Public Sub New(ByVal modelState As ModelStateDictionary)
        MyBase.New(modelState, New EntityFrameworkBlogRepository())
    End Sub

    Public Sub New(ByVal modelState As ModelStateDictionary, ByVal blogRepository As BlogRepositoryBase)
        MyBase.New(modelState, blogRepository)
    End Sub



    Public Overrides Function GetBlogEntry(ByVal year As Integer, ByVal month As Integer, ByVal day As Integer, ByVal name As String) As BlogEntry
        Return _blogRepository.GetBlogEntry(year, month, day, name)
    End Function

    Public Overrides Function GetBlogEntry(ByVal id As Integer) As BlogEntry
        Return _blogRepository.GetBlogEntry(id)
    End Function

    Public Overrides Function ListBlogEntries(ByVal page As Integer?) As PagedList(Of BlogEntry)
        Return _blogRepository.ListBlogEntries(page, Nothing, Nothing, Nothing)
    End Function


    Public Overrides Function ListBlogEntries(ByVal page As Integer?, ByVal year As Integer?, ByVal month As Integer?, ByVal day As Integer?) As PagedList(Of BlogEntry)
        Return _blogRepository.ListBlogEntries(page, year, month, day)
    End Function


    Public Overrides Function CreateBlogEntry(ByVal blogEntryToCreate As BlogEntry) As Boolean
        ' validation
        If blogEntryToCreate.Title.Trim().Length = 0 Then
            _modelState.AddModelError("Title", "Title is required.")
        End If
        If blogEntryToCreate.Title.Length > 500 Then
            _modelState.AddModelError("Title", "Title is too long.")
        End If
        If blogEntryToCreate.Text.Trim().Length = 0 Then
            _modelState.AddModelError("Text", "Text is required.")
        End If
        If blogEntryToCreate.Author.Trim().Length = 0 Then
            _modelState.AddModelError("Author", "Author is required.")
        End If
        If blogEntryToCreate.Description.Trim().Length = 0 Then
            _modelState.AddModelError("Description", "Description is required.")
        End If
        If _modelState.IsValid = False Then
            Return False
        End If

        ' Business Rules
        If String.IsNullOrEmpty(blogEntryToCreate.Name) Then
            blogEntryToCreate.Name = blogEntryToCreate.Title
        End If
        blogEntryToCreate.Name = blogEntryToCreate.Name.Replace(" ", "-")
        blogEntryToCreate.Name = Regex.Replace(blogEntryToCreate.Name, "[""$&+,/:;=?@]", String.Empty)

        If blogEntryToCreate.DatePublished = DateTime.MinValue Then
            blogEntryToCreate.DatePublished = DateTime.Now
        End If

        ' Data access
        _blogRepository.CreateBlogEntry(blogEntryToCreate)
        Return True
    End Function

    ' Comment Methods

    Public Overrides Function CreateComment(ByVal commentToCreate As Comment) As Boolean
        ' Validation
        If commentToCreate.Title.Trim().Length = 0 Then
            _modelState.AddModelError("Title", "Title is required.")
        End If
        If commentToCreate.Name.Trim().Length = 0 Then
            _modelState.AddModelError("Name", "Name is required.")
        End If
        If commentToCreate.Text.Trim().Length = 0 Then
            _modelState.AddModelError("Text", "Comment is required.")
        End If
        If _modelState.IsValid = False Then
            Return False
        End If

        ' Business rules
        If commentToCreate.DatePublished = DateTime.MinValue Then
            commentToCreate.DatePublished = DateTime.Now
        End If

        ' Data access
        _blogRepository.CreateComment(commentToCreate)
        Return True
    End Function


    ' Archive Info methods
    Public Overrides Function ListBlogEntriesByMonth() As IList(Of ArchiveInfo)
        Return CType(_blogRepository.ListBlogEntriesByMonth(), IList(Of ArchiveInfo))
    End Function

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter21/CS/UnleashedBlog/UnleashedBlog/Models/BlogService.cs
using System;
using System.Text.RegularExpressions;
using System.Web.Mvc;
using Paging;
using UnleashedBlog.Models.EntityFramework;
using System.Collections.Generic;

namespace UnleashedBlog.Models
{
    public class BlogService : BlogServiceBase
    {

        public BlogService(ModelStateDictionary modelState)
            : base(modelState, new EntityFrameworkBlogRepository()) { }

        public BlogService(ModelStateDictionary modelState, BlogRepositoryBase blogRepository)
            : base(modelState, blogRepository) { }



        public override BlogEntry GetBlogEntry(int year, int month, int day, string name)
        {
            return _blogRepository.GetBlogEntry(year, month, day, name);
        }

        public override BlogEntry GetBlogEntry(int id)
        {
            return _blogRepository.GetBlogEntry(id);
        }

        public override PagedList<BlogEntry> ListBlogEntries(int? page)
        {
            return _blogRepository.ListBlogEntries(page, null, null, null);
        }


        public override PagedList<BlogEntry> ListBlogEntries(int? page, int? year, int? month, int? day)
        {
            return _blogRepository.ListBlogEntries(page, year, month, day);
        }


        public override bool CreateBlogEntry(BlogEntry blogEntryToCreate)
        {
            // validation
            if (blogEntryToCreate.Title.Trim().Length == 0)
                _modelState.AddModelError("Title", "Title is required.");
            if (blogEntryToCreate.Title.Length > 500)
                _modelState.AddModelError("Title", "Title is too long.");
            if (blogEntryToCreate.Text.Trim().Length == 0)
                _modelState.AddModelError("Text", "Text is required.");
            if (blogEntryToCreate.Author.Trim().Length == 0)
                _modelState.AddModelError("Author", "Author is required.");
            if (blogEntryToCreate.Description.Trim().Length == 0)
                _modelState.AddModelError("Description", "Description is required.");
            if (_modelState.IsValid == false)
                return false;

            // Business Rules
            if (String.IsNullOrEmpty(blogEntryToCreate.Name))
                blogEntryToCreate.Name = blogEntryToCreate.Title;
            blogEntryToCreate.Name = blogEntryToCreate.Name.Replace(" ", "-");
            blogEntryToCreate.Name = Regex.Replace(blogEntryToCreate.Name, "[\"$&+,/:;=?@]", string.Empty);

            if (blogEntryToCreate.DatePublished == DateTime.MinValue)
                blogEntryToCreate.DatePublished = DateTime.Now;

            // Data access
            _blogRepository.CreateBlogEntry(blogEntryToCreate);
            return true;
        }

        // Comment Methods

        public override bool CreateComment(Comment commentToCreate)
        {
            // Validation
            if (commentToCreate.Title.Trim().Length == 0)
                _modelState.AddModelError("Title", "Title is required.");
            if (commentToCreate.Name.Trim().Length == 0)
                _modelState.AddModelError("Name", "Name is required.");
            if (commentToCreate.Text.Trim().Length == 0)
                _modelState.AddModelError("Text", "Comment is required.");
            if (_modelState.IsValid == false)
                return false;

            // Business rules
            if (commentToCreate.DatePublished == DateTime.MinValue)
                commentToCreate.DatePublished = DateTime.Now;

            // Data access
            _blogRepository.CreateComment(commentToCreate);
            return true;
        }


        // Archive Info methods
        public override IList<ArchiveInfo> ListBlogEntriesByMonth()
        {
            return (IList<ArchiveInfo>)_blogRepository.ListBlogEntriesByMonth();
        }

    }
}
